src/lib/gossip_net/libp2p.ml

```diff
@@ -0,0 +1,378 @@
+                            , Some
+                                ( "Handshake error: $exn"
+                                , [("exn", `String (Exn.to_string exn))] ) ))
+                    in
+                    Or_error.error_string "handshake error" )) )
+        >>= function
+        | Ok (Ok result) ->
+            (* call succeeded, result is valid *)
+            Deferred.return (Ok result)
+        | Ok (Error err) -> (
+            (* call succeeded, result is an error *)
+            Logger.error t.config.logger ~module_:__MODULE__ ~location:__LOC__
+              "RPC call error: $error, same error in machine format: \
+               $machine_error"
+              ~metadata:
+                [ ("error", `String (Error.to_string_hum err))
+                ; ("machine_error", `String (Error.to_string_mach err)) ] ;
+            match (Error.to_exn err, Error.sexp_of_t err) with
+            | ( _
+              , Sexp.List
+                  [ Sexp.List
+                      [ Sexp.Atom "rpc_error"
+                      ; Sexp.List [Sexp.Atom "Connection_closed"; _] ]
+                  ; _connection_description
+                  ; _rpc_tag
+                  ; _rpc_version ] ) ->
+                let%map () =
+                  Trust_system.(
+                    record t.config.trust_system t.config.logger addr
+                      Actions.
+                        ( Outgoing_connection_error
+                        , Some ("Closed connection", []) ))
+                in
+                Error err
+            | _ ->
+                let%map () =
+                  Trust_system.(
+                    record t.config.trust_system t.config.logger addr
+                      Actions.
+                        ( Violated_protocol
+                        , Some
+                            ( "RPC call failed, reason: $exn"
+                            , [("exn", `String (Error.to_string_hum err))] ) ))
+                in
+                Error err )
+        | Error monitor_exn -> (
+            (* call itself failed *)
+            (* TODO: learn what other exceptions are raised here *)
+            let exn = Monitor.extract_exn monitor_exn in
+            match exn with
+            | _ ->
+                Logger.error t.config.logger ~module_:__MODULE__
+                  ~location:__LOC__ "RPC call raised an exception: $exn"
+                  ~metadata:[("exn", `String (Exn.to_string exn))] ;
+                Deferred.return (Or_error.of_exn exn) )
+      in
+      call ()
+
+    let try_call_rpc : type q r.
+           t
+        -> Unix.Inet_addr.t
+        -> _
+        -> _
+        -> (q, r) rpc
+        -> q
+        -> r Deferred.Or_error.t =
+     fun t peer rd wr rpc query ->
+      let (module Impl) = implementation_of_rpc rpc in
+      try_call_rpc_with_dispatch t peer rd wr Impl.dispatch_multi query
+
+    let query_peer t (peer : Peer.Id.t) rpc rpc_input =
```

Comment:
Should we be reusing streams? Don't need to do that optimization now

Reply:
streams are pretty lightweight, I think that'd just move a session cookie from libp2p into our code.